zebrad/src/components/mempool.rs

```diff
@@ -87,48 +104,120 @@ pub struct Mempool {
     /// Allows the detection of chain tip resets.
     #[allow(dead_code)]
     chain_tip_change: ChainTipChange,
+
+    /// Handle to the outbound service.
+    /// Used to construct the transaction downloader.
+    outbound_service: OutboundService,
+
+    /// Handle to the state service.
+    /// Used to construct the transaction downloader.
+    state_service: StateService,
+
+    /// Handle to the transaction verifier service.
+    /// Used to construct the transaction downloader.
+    tx_verifier_service: TxVerifierService,
 }
 
 impl Mempool {
     #[allow(dead_code)]
     pub(crate) fn new(
         _network: Network,
-        outbound: Outbound,
-        state: State,
-        tx_verifier: TxVerifier,
+        outbound_service: OutboundService,
+        state_service: StateService,
+        tx_verifier_service: TxVerifierService,
         sync_status: SyncStatus,
         chain_tip_change: ChainTipChange,
     ) -> Self {
-        let tx_downloads = Box::pin(TxDownloads::new(
-            Timeout::new(outbound, TRANSACTION_DOWNLOAD_TIMEOUT),
-            Timeout::new(tx_verifier, TRANSACTION_VERIFY_TIMEOUT),
-            state,
-        ));
-
         Mempool {
-            storage: Default::default(),
-            tx_downloads,
+            state: State::Disabled,
             sync_status,
             chain_tip_change,
+            outbound_service,
+            state_service,
+            tx_verifier_service,
+        }
+    }
+
+    /// Update the mempool state (enabled / disabled) depending on how close to
+    /// the tip is the synchronization, including side effects to state changes.
+    fn update_state(&mut self) {
+        // Update enabled / disabled state
+        let is_close_to_tip = self.sync_status.is_close_to_tip();
+        if is_close_to_tip && matches!(self.state, State::Disabled) {
+            let tx_downloads = Box::pin(TxDownloads::new(
+                Timeout::new(self.outbound_service.clone(), TRANSACTION_DOWNLOAD_TIMEOUT),
+                Timeout::new(self.tx_verifier_service.clone(), TRANSACTION_VERIFY_TIMEOUT),
+                self.state_service.clone(),
+            ));
+            self.state = State::Enabled {
+                storage: Default::default(),
+                tx_downloads,
+            };
+        } else if !is_close_to_tip && matches!(self.state, State::Enabled { .. }) {
+            self.state = State::Disabled
+        }
+    }
+
+    /// Return wether the mempool is enabled or not.
+    #[allow(dead_code)]
+    pub fn enabled(&self) -> bool {
```

Comment:
Naming nit: `is_` and `has_` make boolean methods into questions. This helps distinguish `is_enabled` from `enable`.

```suggestion
    pub fn is_enabled(&self) -> bool {
```

Reply:
Done in fe354f51e